packages/editor/src/systems/RenderMonitorSystem.tsx

```diff
@@ -0,0 +1,134 @@
+import { getMutableState, useHookstate } from '@ir-engine/hyperflux'
+
+import { useQuery } from '@ir-engine/ecs'
+import { SourceComponent } from '@ir-engine/engine/src/scene/components/SourceComponent'
+import { RenderInfoState } from '@ir-engine/spatial/src/renderer/RenderInfoSystem'
+import { VisibleComponent } from '@ir-engine/spatial/src/renderer/components/VisibleComponent'
+import { LightTagComponent } from '@ir-engine/spatial/src/renderer/components/lights/LightTagComponent'
+import { ResourceState, ResourceType } from '@ir-engine/spatial/src/resources/ResourceState'
+import { useTranslation } from 'react-i18next'
+import { EditorWarningState } from '../services/EditorWarningServices'
+
+type SceneComplexityParams = {
+  vertices: number
+  triangles: number
+  texturesMB: number
+  lights: number
+  drawCalls: number
+  shaderComplexity: number
+}
+
+function calculateSceneComplexity(params: SceneComplexityParams): number {
+  const complexity =
+    SceneComplexityWeights.verticesWeight * params.vertices +
+    SceneComplexityWeights.trianglesWeight * params.triangles +
+    SceneComplexityWeights.texturesMBWeight * params.texturesMB +
+    SceneComplexityWeights.lightsWeight * params.lights +
+    SceneComplexityWeights.drawCallsWeight * params.drawCalls +
+    SceneComplexityWeights.shaderComplexityWeight * params.shaderComplexity
+
+  return complexity
+}
+
+// count number of lines in vertex and fragment shaders
+function getShaderComplexity(resources: Record<string, any>): number {
+  const countLines = (code?: string): number => (code || '').split('\n').length
+  const totalComplexity = Object.values(resources)
+    .filter(
+      (resource) =>
+        resource.type === ResourceType.Material &&
+        (resource.asset?.isShaderMaterial || resource.asset?.isMeshStandardMaterial)
+    )
+    .map((resource) => resource.asset)
+    .reduce(
+      (total, material) => {
+        total.vertexInstructions += countLines(material.vertexShader)
+        total.fragmentInstructions += countLines(material.fragmentShader)
+        return total
+      },
+      { vertexInstructions: 0, fragmentInstructions: 0 }
+    )
+
+  return totalComplexity.vertexInstructions + totalComplexity.fragmentInstructions
+}
+
+export const RenderMonitorSystem = defineSystem({
+  uuid: 'ee.editor.RenderMonitorSystem',
+  insert: { after: PresentationSystemGroup },
+  reactor: () => {
+    const { t } = useTranslation()
+
+    const sceneComplexityScore = useHookstate(0)
+
+    const renderInfoState = useHookstate(getMutableState(RenderInfoState))
+    const resourceState = useHookstate(getMutableState(ResourceState))
+    const lightQuery = useQuery([LightTagComponent, VisibleComponent, SourceComponent])
+
+    useEffect(() => {
+      const params = {
+        vertices: resourceState.totalVertexCount.value,
+        triangles: renderInfoState.info.triangles.value,
```

Comment:
I'm not sure that we want to use the render info triangle count here. The reason for storing our own vertex count is because not all geometry should be considered part of the scene budget (Things like transform gizmos and debug meshes).

Reply:
yes, we need to iterate from the loaded scene gltf entity and only include deeply nested entities that belong to the scene

Reply:
roger that